unstructured/ingest/v2/processes/connectors/pinecone.py

```diff
@@ -0,0 +1,181 @@
+
+        pc = Pinecone(
+            api_key=self.access_config.api_key,
+            source_tag=f"unstructured=={unstructured_version}",
+        )
+
+        index = pc.Index(self.index_name)
+        logger.debug(f"Connected to index: {pc.describe_index(self.index_name)}")
+        return index
+
+
+@dataclass
+class PineconeUploadStagerConfig(UploadStagerConfig):
+    pass
+
+
+@dataclass
+class PineconeUploaderConfig(UploaderConfig):
+    batch_size: int = 100
+    num_of_processes: int = 4
+
+
+@dataclass
+class PineconeUploadStager(UploadStager):
+    upload_stager_config: PineconeUploadStagerConfig = field(
+        default_factory=lambda: PineconeUploadStagerConfig()
+    )
+
+    @staticmethod
+    def conform_dict(element_dict: dict) -> dict:
+        # While flatten_dict enables indexing on various fields,
+        # element_serialized enables easily reloading the element object to memory.
+        # element_serialized is formed without text/embeddings to avoid data bloating.
+        return {
+            "id": str(uuid.uuid4()),
+            "values": element_dict.pop("embeddings", None),
+            "metadata": {
+                "text": element_dict.pop("text", None),
+                "element_serialized": json.dumps(element_dict),
+                **flatten_dict(
+                    element_dict,
+                    separator="-",
+                    flatten_lists=True,
+                    remove_none=True,
+                ),
+            },
+        }
+
+    def run(
+        self,
+        elements_filepath: Path,
+        output_dir: Path,
+        output_filename: str,
+        **kwargs: Any,
+    ) -> Path:
+        with open(elements_filepath) as elements_file:
+            elements_contents = json.load(elements_file)
+
+        conformed_elements = [
+            self.conform_dict(element_dict=element) for element in elements_contents
+        ]
+
+        output_path = Path(output_dir) / Path(f"{output_filename}.json")
+        with open(output_path, "w") as output_file:
+            json.dump(conformed_elements, output_file)
+        return output_path
+
+
+@dataclass
+class PineconeUploader(Uploader):
+    upload_config: PineconeUploaderConfig
+    connection_config: PineconeConnectionConfig
+    connector_type: str = CONNECTOR_TYPE
+
+    @DestinationConnectionError.wrap
+    def check_connection(self):
+        _ = self.connection_config.create_index()
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        try:
+            index = self.connection_config.create_index()
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.PineconeApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
```

Comment:
Couple comments:
* Import the exception being caught directly
* If caught, raise `DestinationConnectionError` directly rather than wrapping it

```suggestion
    @requires_dependencies(["pinecone"], extras="pinecone")
    def upsert_batch(self, batch):
        from pinecone.core.client.exceptions import PineconeApiException

        try:
            index = self.connection_config.create_index()
            response = index.upsert(batch)
        except PineconeApiException as api_error:
            raise DestinationConnectionError(f"http error: {api_error}") from api_error
        logger.debug(f"results: {response}")
```

Reply:
[import exception directly, raise exception directly rather than wrapp…](https://github.com/Unstructured-IO/unstructured/pull/3286/commits/e89dcf60ffcfd4a3f0ce218e0ef45020073264d8)

Reply:
Btw, why do we prefer raising directly in this case?